In [1]:
# Load and activate the SQL extension to allow us to execute SQL in a Jupyter notebook.
%load_ext sql

In [3]:
# Establish a connection to the local database using the '%sql' magic command.
%sql mysql+pymysql://root:EsLAMmohamed712002@127.0.0.1:3306/md_water_services

'Connected: root@md_water_services'

# Part 1: Gaining insights from our Database

In [8]:
%%sql
-- feel the database

SELECT
    *
FROM
    data_dictionary;

 * mysql+pymysql://root:***@127.0.0.1:3306/md_water_services
49 rows affected.


[('employee', 'assigned_employee_id', 'Unique ID assigned to each employee', 'INT', 'visits'),
 ('employee', 'employee_name', 'Name of the employee', 'VARCHAR(255)', ''),
 ('employee', 'phone_number', 'Contact number of the employee', 'VARCHAR(15)', ''),
 ('employee', 'email', 'Email address of the employee', 'VARCHAR(255)', ''),
 ('employee', 'address', 'Residential address of the employee', 'VARCHAR(255)', ''),
 ('employee', 'town_name', 'Name of the town where the employee resides', 'VARCHAR(255)', ''),
 ('employee', 'province_name', 'Name of the province where the employee resides', 'VARCHAR(255)', ''),
 ('employee', 'position', 'Position or job title of the employee', 'VARCHAR(255)', ''),
 ('visits', 'record_id', 'Unique ID assigned to each visit', 'int', 'water_quality, water_source'),
 ('visits', 'location_id', 'ID of the location visited', 'varchar(255)', 'location'),
 ('visits', 'source_id', 'ID of the water source visited', 'varchar(510)', 'well_pollution'),
 ('visits', 'time_of_record', 'Date and time of the visit', 'datetime', ''),
 ('visits', 'visit_count', 'Number of visits made to this location', 'int', ''),
 ('visits', 'time_in_queue', 'Time spent by people waiting for water in a queue at the location', 'int', ''),
 ('visits', 'assigned_employee_id', 'ID of the employee who visited the location', 'int', 'employee'),
 ('water_quality', 'record_id', 'Unique ID assigned to each record', 'int', ''),
 ('water_quality', 'subjective_quality_score', 'Score representing the subjective quality of the water source', 'int', ''),
 ('water_quality', 'visit_count', 'Number of visits made for data collection', 'int', ''),
 ('water_source', 'source_id', 'Unique ID assigned to each water source', 'VARCHAR(510)', 'visits'),
 ('water_source', 'type_of_water_source', 'Type or category of the water source. Can be : tap_in_home, tap_in_home_broken, well, shared_tap, river', 'VARCHAR(255)', ''),
 ('water_source', 'number_of_people_served', 'Number of people served by this water source', 'INT', ''),
 ('well_pollution', 'source_id', 'ID of the water source being tested for pollution', 'varchar(258)', 'visits'),
 ('well_pollution', 'date', 'Date of the pollution test', 'datetime', ''),
 ('well_pollution', 'description', 'Description of the pollution test', 'varchar(255)', ''),
 ('well_pollution', 'pollutant_ppm', 'Result of the pollution test in parts per million', 'float', ''),
 ('well_pollution', 'biological', 'Biological contamination level', 'float', ''),
 ('well_pollution', 'results', "Result of the pollution test. Can be 'dirty', 'biologically contaminated', or 'clean'", 'varchar(255)', ''),
 ('location', 'location_id', 'Unique ID assigned to each location', 'varchar(255)', 'visits'),
 ('location', 'address', 'Address of the location', 'varchar(255)', ''),
 ('location', 'province_name', 'Name of the province where the location is situated', 'varchar(255)', ''),
 ('location', 'town_name', 'Name of the town where the location is situated', 'varchar(255)', ''),
 ('location', 'location_type', 'Type or category of the location', 'varchar(255)', ''),
 ('global_water_access', 'name', 'The country or area name', 'varchar(255)', ''),
 ('global_water_access', 'region', 'Geographical region', 'varchar(255)', ''),
 ('global_water_access', 'year', 'Year of the data record', 'int', ''),
 ('global_water_access', 'pop_n', 'The national population size estimate in thousands', 'float', ''),
 ('global_water_access', 'pop_u', 'The urban population share estimate in percentage points (%)', 'float', ''),
 ('global_water_access', 'wat_bas_n', 'The estimated national share of people with at least basic service (%)', 'float', ''),
 ('global_water_access', 'wat_lim_n', 'The estimated national share of people with limited service (%)', 'float', ''),
 ('global_water_access', 'wat_unimp_n', 'The estimated national share of people with unimproved service (%)', 'float', ''),
 ('global_water_access', 'wat_sur_n', 'The estimated national share of people with surfa

In [277]:
%%sql
-- update employees email

# UPDATE
#     employee
# SET
#     email = LOWER(CONCAT(REPLACE(employee_name, ' ', '.'), '@ndogowater.gov'))

SELECT
    email
FROM
    employee
LIMIT 5;

 * mysql+pymysql://root:***@127.0.0.1:3306/md_water_services
5 rows affected.


[('amara.jengo@ndogowater.gov',),
 ('bello.azibo@ndogowater.gov',),
 ('bakari.iniko@ndogowater.gov',),
 ('malachi.mavuso@ndogowater.gov',),
 ('cheche.buhle@ndogowater.gov',)]

In [27]:
%%sql

-- update employees phone_number

# UPDATE
#      employee
# SET
#      Phone_number = TRIM(Phone_number)

SELECT
    LENGTH(Phone_number)
FROM
    employee
LIMIT 5;

 * mysql+pymysql://root:***@127.0.0.1:3306/md_water_services
5 rows affected.


[(12,), (12,), (12,), (12,), (12,)]

In [31]:
%%sql
-- know where our employees live

SELECT
    town_name,
    COUNT(employee_name) AS employees_residens
FROM
    employee
GROUP BY
    town_name;

 * mysql+pymysql://root:***@127.0.0.1:3306/md_water_services
9 rows affected.


[('Ilanga', 3),
 ('Rural', 29),
 ('Lusaka', 4),
 ('Zanzibar', 4),
 ('Dahabu', 6),
 ('Kintampo', 1),
 ('Harare', 5),
 ('Yaounde', 1),
 ('Serowe', 3)]

In [44]:
%%sql
--  finding the top 3 employees with the highest number of locations visited.

SELECT
    assigned_employee_id,
    COUNT(visit_count) AS number_of_visits
FROM
    visits
GROUP BY
    assigned_employee_id
ORDER BY
    COUNT(visit_count) DESC
LIMIT 3;

 * mysql+pymysql://root:***@127.0.0.1:3306/md_water_services
3 rows affected.


[(1, 3708), (30, 3676), (34, 3539)]

In [48]:
%%sql
-- hounouring the workers

SELECT
    assigned_employee_id,
    employee_name,
    phone_number,
    email
FROM
    employee
WHERE
    assigned_employee_id IN(1, 30, 34);

 * mysql+pymysql://root:***@127.0.0.1:3306/md_water_services
3 rows affected.


[(1, 'Bello Azibo', '+99643864786', 'bello.azibo@ndogowater.gov'),
 (30, 'Pili Zola', '+99822478933', 'pili.zola@ndogowater.gov'),
 (34, 'Rudo Imani', '+99046972648', 'rudo.imani@ndogowater.gov')]

In [95]:
%%sql
-- Analysing locations

SELECT
    province_name,
    town_name,
    COUNT(location_id) AS records_per_town
FROM
    location
GROUP BY
    town_name, province_name
ORDER BY
    province_name, COUNT(location_id) DESC;

 * mysql+pymysql://root:***@127.0.0.1:3306/md_water_services
31 rows affected.


[('Akatsi', 'Rural', 6290),
 ('Akatsi', 'Lusaka', 1070),
 ('Akatsi', 'Harare', 800),
 ('Akatsi', 'Kintampo', 780),
 ('Amanzi', 'Rural', 3100),
 ('Amanzi', 'Asmara', 930),
 ('Amanzi', 'Dahabu', 930),
 ('Amanzi', 'Amina', 670),
 ('Amanzi', 'Pwani', 520),
 ('Amanzi', 'Abidjan', 400),
 ('Amanzi', 'Bello', 400),
 ('Hawassa', 'Rural', 3900),
 ('Hawassa', 'Serowe', 480),
 ('Hawassa', 'Amina', 420),
 ('Hawassa', 'Deka', 350),
 ('Hawassa', 'Djenne', 310),
 ('Hawassa', 'Zanzibar', 300),
 ('Hawassa', 'Yaounde', 270),
 ('Kilimani', 'Rural', 5440),
 ('Kilimani', 'Mrembo', 990),
 ('Kilimani', 'Harare', 850),
 ('Kilimani', 'Isiqalo', 770),
 ('Kilimani', 'Zuri', 750),
 ('Kilimani', 'Amara', 710),
 ('Sokoto', 'Rural', 5010),
 ('Sokoto', 'Ilanga', 780),
 ('Sokoto', 'Majengo', 620),
 ('Sokoto', 'Kofi', 510),
 ('Sokoto', 'Bahari', 470),
 ('Sokoto', 'Marang', 460),
 ('Sokoto', 'Cheche', 370)]

In [102]:
%%sql
-- Analysing locations

SELECT
    location_type,
    COUNT(location_id) AS records_per_type
FROM
    location
GROUP BY
    location_type
ORDER BY
    records_per_type DESC;

 * mysql+pymysql://root:***@127.0.0.1:3306/md_water_services
2 rows affected.


[('Rural', 23740), ('Urban', 15910)]

# Some insights from the location table:
1. Our entire country was properly canvassed, and our dataset represents the situation on the ground.
2. 60% of our water sources are in rural communities across Maji Ndogo. We need to keep this in mind when we make decisions.

In [106]:
%%sql
-- How many people did we survey in total?

SELECT
    SUM(number_of_people_served) AS total_people_served
FROM
    water_source

 * mysql+pymysql://root:***@127.0.0.1:3306/md_water_services
1 rows affected.


[(Decimal('27628140'),)]

In [113]:
%%sql
-- How many wells, taps and rivers are there?

SELECT
    type_of_water_source,
    COUNT(number_of_people_served) AS number_of_water_source
FROM
    water_source
GROUP BY
    type_of_water_source;

 * mysql+pymysql://root:***@127.0.0.1:3306/md_water_services
5 rows affected.


[('tap_in_home', 7265),
 ('tap_in_home_broken', 5856),
 ('well', 17383),
 ('shared_tap', 5767),
 ('river', 3379)]

In [117]:
%%sql
-- How many people share particular types of water sources on average?

SELECT
    type_of_water_source,
    ROUND(AVG(number_of_people_served)) AS avg_people_per_source
FROM
    water_source
GROUP BY
    type_of_water_source;

 * mysql+pymysql://root:***@127.0.0.1:3306/md_water_services
5 rows affected.


[('tap_in_home', Decimal('644')),
 ('tap_in_home_broken', Decimal('649')),
 ('well', Decimal('279')),
 ('shared_tap', Decimal('2071')),
 ('river', Decimal('699'))]

In [122]:
%%sql
-- How many people are getting water from each type of source?

SELECT
    type_of_water_source,
    ROUND(SUM(number_of_people_served) / 27628140 * 100) AS population_served
FROM
    water_source
GROUP BY
    type_of_water_source
ORDER BY
    population_served DESC;

 * mysql+pymysql://root:***@127.0.0.1:3306/md_water_services
5 rows affected.


[('shared_tap', Decimal('43')),
 ('well', Decimal('18')),
 ('tap_in_home', Decimal('17')),
 ('tap_in_home_broken', Decimal('14')),
 ('river', Decimal('9'))]

In [173]:
%%sql

SELECT
    source_id,
    type_of_water_source,
    number_of_people_served,
    ROW_NUMBER() OVER(
    PARTITION BY type_of_water_source ORDER BY number_of_people_served DESC) AS rank_by_population
FROM
    water_source
HAVING
    type_of_water_source <> 'tap_in_home'
LIMIT 10;

 * mysql+pymysql://root:***@127.0.0.1:3306/md_water_services
10 rows affected.


[('SoRu36791224', 'river', 998, 1),
 ('SoRu36238224', 'river', 998, 2),
 ('SoRu36880224', 'river', 998, 3),
 ('SoRu37756224', 'river', 998, 4),
 ('SoRu38142224', 'river', 998, 5),
 ('KiRu30353224', 'river', 998, 6),
 ('KiZu31236224', 'river', 998, 7),
 ('KiRu28591224', 'river', 998, 8),
 ('KiMr25030224', 'river', 998, 9),
 ('SoRu35837224', 'river', 998, 10)]

# Some insights from the water source table:
1. 43% of our people are using shared taps in their communities.
2. 31% of people have water infrastructure installed in their homes, but 45% (14/31) of these taps are not working!
3. 18% of people are using wells. But only 4916 out of 17383 are clean = 28%.

In [175]:
%%sql

SELECT
    *
FROM
    visits
LIMIT 10;

 * mysql+pymysql://root:***@127.0.0.1:3306/md_water_services
10 rows affected.


[(0, 'SoIl32582', 'SoIl32582224', datetime.datetime(2021, 1, 1, 9, 10), 1, 15, 12),
 (1, 'KiRu28935', 'KiRu28935224', datetime.datetime(2021, 1, 1, 9, 17), 1, 0, 46),
 (2, 'HaRu19752', 'HaRu19752224', datetime.datetime(2021, 1, 1, 9, 36), 1, 62, 40),
 (3, 'AkLu01628', 'AkLu01628224', datetime.datetime(2021, 1, 1, 9, 53), 1, 0, 1),
 (4, 'AkRu03357', 'AkRu03357224', datetime.datetime(2021, 1, 1, 10, 11), 1, 28, 14),
 (5, 'KiRu29315', 'KiRu29315224', datetime.datetime(2021, 1, 1, 10, 17), 1, 9, 40),
 (6, 'AkRu05234', 'AkRu05234224', datetime.datetime(2021, 1, 1, 10, 18), 1, 0, 30),
 (7, 'KiRu28520', 'KiRu28520224', datetime.datetime(2021, 1, 1, 10, 28), 1, 0, 34),
 (8, 'HaZa21742', 'HaZa21742224', datetime.datetime(2021, 1, 1, 10, 37), 1, 0, 6),
 (9, 'AmDa12214', 'AmDa12214224', datetime.datetime(2021, 1, 1, 10, 58), 1, 0, 36)]

In [196]:
%%sql
-- How long did the survey take?

SELECT
    DATEDIFF(MAX(time_of_record), MIN(time_of_record)) AS number_of_days
FROM
    visits;

 * mysql+pymysql://root:***@127.0.0.1:3306/md_water_services
1 rows affected.


[(924,)]

In [204]:
%%sql
-- What is the average total queue time for water?

SELECT
    ROUND(AVG(NULLIF(time_in_queue, 0))) AS avg_qeue_time
FROM
    visits;

 * mysql+pymysql://root:***@127.0.0.1:3306/md_water_services
1 rows affected.


[(Decimal('123'),)]

In [211]:
%%sql
-- What is the average queue time on different days?

SELECT
    DAYNAME(time_of_record) AS day_of_week,
    ROUND(AVG(NULLIF(time_in_queue, 0))) AS avg_qeue_time
FROM
    visits
GROUP BY
    day_of_week
ORDER BY
    avg_qeue_time DESC;

 * mysql+pymysql://root:***@127.0.0.1:3306/md_water_services
7 rows affected.


[('Saturday', Decimal('246')),
 ('Monday', Decimal('137')),
 ('Friday', Decimal('120')),
 ('Tuesday', Decimal('108')),
 ('Thursday', Decimal('105')),
 ('Wednesday', Decimal('97')),
 ('Sunday', Decimal('82'))]

In [215]:
%%sql
-- What is the average queue time on different hours?

SELECT
    TIME_FORMAT(TIME(time_of_record), '%H:00') AS hour_of_week,
    ROUND(AVG(NULLIF(time_in_queue, 0))) AS avg_qeue_time
FROM
    visits
GROUP BY
    hour_of_week
ORDER BY
    hour_of_week;

 * mysql+pymysql://root:***@127.0.0.1:3306/md_water_services
14 rows affected.


[('06:00', Decimal('149')),
 ('07:00', Decimal('149')),
 ('08:00', Decimal('149')),
 ('09:00', Decimal('118')),
 ('10:00', Decimal('114')),
 ('11:00', Decimal('111')),
 ('12:00', Decimal('112')),
 ('13:00', Decimal('115')),
 ('14:00', Decimal('114')),
 ('15:00', Decimal('114')),
 ('16:00', Decimal('114')),
 ('17:00', Decimal('149')),
 ('18:00', Decimal('147')),
 ('19:00', Decimal('168'))]

In [219]:
%%sql
-- What is the average queue time on different hours on different days?

SELECT
    TIME_FORMAT(TIME(time_of_record), '%H:00') AS hour_of_week,
    ROUND(AVG(
    CASE
        WHEN DAYNAME(time_of_record) = 'Saturday' THEN time_in_queue
        ELSE NULL
    END), 0) AS Saturday,
    ROUND(AVG(
    CASE
        WHEN DAYNAME(time_of_record) = 'Sunday' THEN time_in_queue
        ELSE NULL
    END), 0) AS Sunday,
    ROUND(AVG(
    CASE
        WHEN DAYNAME(time_of_record) = 'Monday' THEN time_in_queue
        ELSE NULL
    END), 0) AS Monday,
    ROUND(AVG(
    CASE
        WHEN DAYNAME(time_of_record) = 'Tuesday' THEN time_in_queue
        ELSE NULL
    END), 0) AS  Tuesday,
    ROUND(AVG(
    CASE
        WHEN DAYNAME(time_of_record) = 'Wednesday' THEN time_in_queue
        ELSE NULL
    END), 0) AS Wednesday,
    ROUND(AVG(
    CASE
        WHEN DAYNAME(time_of_record) = 'Thursday' THEN time_in_queue
        ELSE NULL
    END), 0) AS Thursday,
    ROUND(AVG(
    CASE
        WHEN DAYNAME(time_of_record) = 'Friday' THEN time_in_queue
        ELSE NULL
    END), 0) AS Friday
FROM
    visits
WHERE
    time_in_queue <> 0
GROUP BY
    hour_of_week
ORDER BY
    hour_of_week;

 * mysql+pymysql://root:***@127.0.0.1:3306/md_water_services
14 rows affected.


[('06:00', Decimal('247'), Decimal('79'), Decimal('190'), Decimal('134'), Decimal('112'), Decimal('134'), Decimal('153')),
 ('07:00', Decimal('247'), Decimal('82'), Decimal('186'), Decimal('128'), Decimal('111'), Decimal('139'), Decimal('156')),
 ('08:00', Decimal('247'), Decimal('86'), Decimal('183'), Decimal('130'), Decimal('119'), Decimal('129'), Decimal('153')),
 ('09:00', Decimal('252'), Decimal('84'), Decimal('127'), Decimal('105'), Decimal('94'), Decimal('99'), Decimal('107')),
 ('10:00', Decimal('259'), Decimal('83'), Decimal('119'), Decimal('99'), Decimal('89'), Decimal('95'), Decimal('112')),
 ('11:00', Decimal('236'), Decimal('78'), Decimal('115'), Decimal('102'), Decimal('86'), Decimal('99'), Decimal('104')),
 ('12:00', Decimal('239'), Decimal('78'), Decimal('115'), Decimal('97'), Decimal('88'), Decimal('96'), Decimal('109')),
 ('13:00', Decimal('242'), Decimal('81'), Decimal('122'), Decimal('97'), Decimal('98'), Decimal('101'), Decimal('115')),
 ('14:00', Decimal('244'), Decimal('83'), Decimal('127'), Decimal('104'), Decimal('92'), Decimal('96'), Decimal('110')),
 ('15:00', Decimal('248'), Decimal('83'), Decimal('126'), Decimal('104'), Decimal('88'), Decimal('92'), Decimal('110')),
 ('16:00', Decimal('251'), Decimal('83'), Decimal('127'), Decimal('99'), Decimal('90'), Decimal('99'), Decimal('109')),
 ('17:00', Decimal('251'), Decimal('79'), Decimal('181'), Decimal('135'), Decimal('121'), Decimal('129'), Decimal('151')),
 ('18:00', Decimal('240'), Decimal('80'), Decimal('174'), Decimal('122'), Decimal('113'), Decimal('132'), Decimal('158')),
 ('19:00', Decimal('282'), Decimal('127'), Decimal('159'), Decimal('145'), Decimal('176'), Decimal('137'), Decimal('103'))]

# Some insights from the visits table:
1. Our citizens often face long wait times for water, averaging more than 120 minutes.
2. Queues are very long on Saturdays.
3. Queues are longer in the mornings and evenings.
4. Wednesdays and Sundays have the shortest queues.

# Part 2: Answering questions

1. Which SQL query will produce the date format "DD Month YYYY" from the time_of_record column in the visits table, as a single column?

SELECT CONCAT(day(time_of_record), " ", monthname(time_of_record), " ", year(time_of_record)) FROM visits;

In [274]:
%%sql
SELECT
    CONCAT(day(time_of_record), " ", monthname(time_of_record), " ", year(time_of_record)) AS time_record
FROM
    visits
LIMIT 10;

 * mysql+pymysql://root:***@127.0.0.1:3306/md_water_services
10 rows affected.


[('1 January 2021',),
 ('1 January 2021',),
 ('1 January 2021',),
 ('1 January 2021',),
 ('1 January 2021',),
 ('1 January 2021',),
 ('1 January 2021',),
 ('1 January 2021',),
 ('1 January 2021',),
 ('1 January 2021',)]

2. You are working with an SQL query designed to calculate the Annual Rate of Change (ARC) for basic rural water services:

    SELECT
    name,
    wat_bas_r - LAG(wat_bas_r) OVER (PARTITION BY (a) ORDER BY (b)) 
    FROM 
    global_water_access
    ORDER BY
    name;

    To accomplish this task, what should you use for placeholders (a) and (b)?

Use name for grouping and year for sorting.

In [250]:
%%sql

SELECT
    name,
    wat_bas_r - LAG(wat_bas_r) OVER (PARTITION BY name ORDER BY year) AS arc
FROM 
    global_water_access
ORDER BY
    name
LIMIT 5;

 * mysql+pymysql://root:***@127.0.0.1:3306/md_water_services
5 rows affected.


[('Afghanistan', None),
 ('Afghanistan', 13.339405059814453),
 ('Albania', None),
 ('Albania', 3.4641036987304688),
 ('Algeria', None)]

3. What are the names of the two worst-performing employees who visited the fewest sites, and how many sites did the worst-performing employee visit? 

Lesedi Kofi, Kunto Asha (15)

In [258]:
%%sql

SELECT
    assigned_employee_id,
    COUNT(visit_count) AS number_of_visits
FROM
    visits
GROUP BY
    assigned_employee_id
ORDER BY
    COUNT(visit_count)
LIMIT 3;

 * mysql+pymysql://root:***@127.0.0.1:3306/md_water_services
3 rows affected.


[(20, 15), (22, 143), (44, 218)]

4. What does the following query do?

    SELECT 
        location_id,
        time_in_queue,
        AVG(time_in_queue) OVER (PARTITION BY location_id ORDER BY visit_count) AS total_avg_queue_time
    FROM 
        visits
    WHERE 
    visit_count > 1 -- Only shared taps were visited > 1
    ORDER BY 
        location_id, time_of_record;

It computes an average queue time for shared taps visited more than once, which is updated each time a source is visited.

In [261]:
%%sql
SELECT
    visit_count,
    location_id,
    time_in_queue,
    AVG(time_in_queue) OVER (PARTITION BY location_id ORDER BY visit_count) AS total_avg_queue_time
FROM 
    visits
WHERE 
visit_count > 1 -- Only shared taps were visited > 1
ORDER BY 
    location_id, time_of_record
LIMIT 20;

 * mysql+pymysql://root:***@127.0.0.1:3306/md_water_services
20 rows affected.


[(2, 'AkHa00036', 113, Decimal('113.0000')),
 (3, 'AkHa00036', 229, Decimal('171.0000')),
 (4, 'AkHa00036', 400, Decimal('247.3333')),
 (5, 'AkHa00036', 120, Decimal('215.5000')),
 (6, 'AkHa00036', 185, Decimal('209.4000')),
 (7, 'AkHa00036', 158, Decimal('200.8333')),
 (8, 'AkHa00036', 226, Decimal('204.4286')),
 (2, 'AkHa00090', 50, Decimal('50.0000')),
 (3, 'AkHa00090', 63, Decimal('56.5000')),
 (4, 'AkHa00090', 12, Decimal('41.6667')),
 (5, 'AkHa00090', 62, Decimal('46.7500')),
 (6, 'AkHa00090', 53, Decimal('48.0000')),
 (7, 'AkHa00090', 34, Decimal('45.6667')),
 (8, 'AkHa00090', 40, Decimal('44.8571')),
 (2, 'AkHa00103', 186, Decimal('186.0000')),
 (3, 'AkHa00103', 88, Decimal('137.0000')),
 (4, 'AkHa00103', 237, Decimal('170.3333')),
 (5, 'AkHa00103', 141, Decimal('163.0000')),
 (6, 'AkHa00103', 183, Decimal('167.0000')),
 (7, 'AkHa00103', 179, Decimal('169.0000'))]

5. One of our employees, Farai Nia, lives at 33 Angelique Kidjo Avenue. What would be the result if we TRIM() her address?

    TRIM('33 Angelique Kidjo Avenue  ')

‘33 Angelique Kidjo Avenue’

In [262]:
%%sql

SELECT
    TRIM('33 Angelique Kidjo Avenue  ');

 * mysql+pymysql://root:***@127.0.0.1:3306/md_water_services
1 rows affected.


[('33 Angelique Kidjo Avenue',)]

6. How many employees live in Dahabu?

6

In [263]:
%%sql

SELECT
    town_name,
    COUNT(employee_name) AS employees_residens
FROM
    employee
GROUP BY
    town_name;

 * mysql+pymysql://root:***@127.0.0.1:3306/md_water_services
9 rows affected.


[('Ilanga', 3),
 ('Rural', 29),
 ('Lusaka', 4),
 ('Zanzibar', 4),
 ('Dahabu', 6),
 ('Kintampo', 1),
 ('Harare', 5),
 ('Yaounde', 1),
 ('Serowe', 3)]

7. How many employees live in Harare, Kilimani?

2

In [279]:
%%sql

SELECT
    province_name,
    town_name,
    COUNT(employee_name) AS employees_residens
FROM
    employee
GROUP BY
    province_name,
    town_name;

 * mysql+pymysql://root:***@127.0.0.1:3306/md_water_services
15 rows affected.


[('Sokoto', 'Ilanga', 2),
 ('Kilimani', 'Rural', 9),
 ('Hawassa', 'Rural', 7),
 ('Akatsi', 'Lusaka', 4),
 ('Akatsi', 'Rural', 5),
 ('Hawassa', 'Zanzibar', 4),
 ('Amanzi', 'Dahabu', 6),
 ('Sokoto', 'Rural', 7),
 ('Akatsi', 'Kintampo', 1),
 ('Akatsi', 'Harare', 3),
 ('Kilimani', 'Ilanga', 1),
 ('Hawassa', 'Yaounde', 1),
 ('Hawassa', 'Serowe', 3),
 ('Kilimani', 'Harare', 2),
 ('Amanzi', 'Rural', 1)]

8. How many people share a well on average?

279

In [264]:
%%sql

SELECT
    type_of_water_source,
    ROUND(AVG(number_of_people_served)) AS avg_people_per_source
FROM
    water_source
GROUP BY
    type_of_water_source;

 * mysql+pymysql://root:***@127.0.0.1:3306/md_water_services
5 rows affected.


[('tap_in_home', Decimal('644')),
 ('tap_in_home_broken', Decimal('649')),
 ('well', Decimal('279')),
 ('shared_tap', Decimal('2071')),
 ('river', Decimal('699'))]

9. Consider the query we used to calculate the total number of people served by each water source:

    SELECT
    type_of_water_source,
    SUM(number_of_people_served) AS population_served
    FROM
    water_source
    GROUP BY
    type_of_water_source
    ORDER BY
    population_served DESC;

    Which of the following lines of code will calculate the total number of people using some sort of tap?

WHERE type_of_water_source LIKE "%tap%"

In [269]:
%%sql

SELECT
    type_of_water_source,
    SUM(number_of_people_served) AS population_served
FROM
    water_source
WHERE
    type_of_water_source LIKE "%tap%"
GROUP BY
    type_of_water_source
ORDER BY
    population_served DESC


 * mysql+pymysql://root:***@127.0.0.1:3306/md_water_services
3 rows affected.


[('shared_tap', Decimal('11945272')),
 ('tap_in_home', Decimal('4678880')),
 ('tap_in_home_broken', Decimal('3799720'))]

10. Use the pivot table we created to answer the following question. What are the average queue times for the following times?

    Saturday from 12:00 to 13:00
    Tuesday from 18:00 to 19:00
    Sunday from 09:00 to 10:00

Saturday: 239, Tuesday: 122, Sunday: 84

In [271]:
%%sql

SELECT
    TIME_FORMAT(TIME(time_of_record), '%H:00') AS hour_of_week,
    ROUND(AVG(
    CASE
        WHEN DAYNAME(time_of_record) = 'Saturday' THEN time_in_queue
        ELSE NULL
    END), 0) AS Saturday,
    ROUND(AVG(
    CASE
        WHEN DAYNAME(time_of_record) = 'Sunday' THEN time_in_queue
        ELSE NULL
    END), 0) AS Sunday,
    ROUND(AVG(
    CASE
        WHEN DAYNAME(time_of_record) = 'Monday' THEN time_in_queue
        ELSE NULL
    END), 0) AS Monday,
    ROUND(AVG(
    CASE
        WHEN DAYNAME(time_of_record) = 'Tuesday' THEN time_in_queue
        ELSE NULL
    END), 0) AS  Tuesday,
    ROUND(AVG(
    CASE
        WHEN DAYNAME(time_of_record) = 'Wednesday' THEN time_in_queue
        ELSE NULL
    END), 0) AS Wednesday,
    ROUND(AVG(
    CASE
        WHEN DAYNAME(time_of_record) = 'Thursday' THEN time_in_queue
        ELSE NULL
    END), 0) AS Thursday,
    ROUND(AVG(
    CASE
        WHEN DAYNAME(time_of_record) = 'Friday' THEN time_in_queue
        ELSE NULL
    END), 0) AS Friday
FROM
    visits
WHERE
    time_in_queue <> 0
GROUP BY
    hour_of_week
ORDER BY
    hour_of_week;

 * mysql+pymysql://root:***@127.0.0.1:3306/md_water_services
14 rows affected.


[('06:00', Decimal('247'), Decimal('79'), Decimal('190'), Decimal('134'), Decimal('112'), Decimal('134'), Decimal('153')),
 ('07:00', Decimal('247'), Decimal('82'), Decimal('186'), Decimal('128'), Decimal('111'), Decimal('139'), Decimal('156')),
 ('08:00', Decimal('247'), Decimal('86'), Decimal('183'), Decimal('130'), Decimal('119'), Decimal('129'), Decimal('153')),
 ('09:00', Decimal('252'), Decimal('84'), Decimal('127'), Decimal('105'), Decimal('94'), Decimal('99'), Decimal('107')),
 ('10:00', Decimal('259'), Decimal('83'), Decimal('119'), Decimal('99'), Decimal('89'), Decimal('95'), Decimal('112')),
 ('11:00', Decimal('236'), Decimal('78'), Decimal('115'), Decimal('102'), Decimal('86'), Decimal('99'), Decimal('104')),
 ('12:00', Decimal('239'), Decimal('78'), Decimal('115'), Decimal('97'), Decimal('88'), Decimal('96'), Decimal('109')),
 ('13:00', Decimal('242'), Decimal('81'), Decimal('122'), Decimal('97'), Decimal('98'), Decimal('101'), Decimal('115')),
 ('14:00', Decimal('244'), Decimal('83'), Decimal('127'), Decimal('104'), Decimal('92'), Decimal('96'), Decimal('110')),
 ('15:00', Decimal('248'), Decimal('83'), Decimal('126'), Decimal('104'), Decimal('88'), Decimal('92'), Decimal('110')),
 ('16:00', Decimal('251'), Decimal('83'), Decimal('127'), Decimal('99'), Decimal('90'), Decimal('99'), Decimal('109')),
 ('17:00', Decimal('251'), Decimal('79'), Decimal('181'), Decimal('135'), Decimal('121'), Decimal('129'), Decimal('151')),
 ('18:00', Decimal('240'), Decimal('80'), Decimal('174'), Decimal('122'), Decimal('113'), Decimal('132'), Decimal('158')),
 ('19:00', Decimal('282'), Decimal('127'), Decimal('159'), Decimal('145'), Decimal('176'), Decimal('137'), Decimal('103'))]